execution/executor/src/db_bootstrapper.rs

```diff
@@ -4,49 +4,51 @@
 #![forbid(unsafe_code)]
 
 use crate::{BlockExecutor, Executor};
-use anyhow::Result;
+use anyhow::{format_err, Result};
 use libra_config::config::NodeConfig;
 use libra_crypto::{hash::PRE_GENESIS_BLOCK_ID, HashValue};
 use libra_logger::prelude::*;
 use libra_temppath::TempPath;
 use libra_types::{
-    ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
-    transaction::Transaction,
-    waypoint::Waypoint,
+    ledger_info::LedgerInfoWithSignatures, transaction::Transaction, waypoint::Waypoint,
 };
 use libra_vm::VMExecutor;
 use libradb::LibraDB;
 use storage_interface::DbReaderWriter;
+use storage_proto::TreeState;
 
-pub fn maybe_bootstrap_db<V: VMExecutor>(db: DbReaderWriter, config: &NodeConfig) -> Result<()> {
-    let startup_info_opt = db.reader.get_startup_info()?;
-    if startup_info_opt.is_some() {
-        return Ok(());
-    }
-
-    let genesis_txn = config
+pub fn bootstrap_db_if_empty<V: VMExecutor>(
+    db: &DbReaderWriter,
+    config: &NodeConfig,
+) -> Result<Option<Waypoint>> {
+    let genesis = config
         .execution
         .genesis
         .as_ref()
         .expect("failed to load genesis txn!")
         .clone();
 
-    execute_and_commit::<V>(db, genesis_txn).map(|_| ())
+    bootstrap_db_if_empty_impl::<V>(db, genesis)
 }
 
-pub fn compute_genesis_waypoint<V: VMExecutor>(genesis: Transaction) -> Result<Waypoint> {
-    let path = TempPath::new();
-    path.create_as_dir().expect("Unable to create directory");
-    let (_, db) = DbReaderWriter::wrap(LibraDB::new(&path.path()));
-    let ledger_info = execute_and_commit::<V>(db, genesis)?;
-    Waypoint::new(&ledger_info)
+fn bootstrap_db_if_empty_impl<V: VMExecutor>(
+    db: &DbReaderWriter,
+    genesis: Transaction,
```

Comment:
here and below

```suggestion
    genesis_txn: Transaction,
```

Reply:
I love you. Hate @davidiw's naming as well. -_,-